src/storage/src/hummock/iterator/mod.rs

```diff
@@ -447,7 +466,14 @@ impl<'a, B: RustIteratorBuilder> HummockIterator for FromRustIterator<'a, B> {
                     assert_eq!(first_key, key.user_key.table_key);
                     match iter.next() {
                         Some((next_key, next_value)) => {
-                            assert_gt!(next_key, first_key);
+                            match Self::Direction::direction() {
+                                DirectionEnum::Forward => {
+                                    assert_gt!(next_key, first_key);
+                                }
+                                DirectionEnum::Backward => {
+                                    assert_lt!(next_key, first_key);
+                                }
```

Comment:
Although it fixes the corner case mentioned [here](https://github.com/risingwavelabs/risingwave/pull/12570#discussion_r1603387103), it introduces a bug for a case that work before:
```
Memtable in epoch2:
 k1

rev iter seek(FullKey(k2, epoch2))

The code will panic in L466 because (k1, epoch2) < (k2, epoch2) but k1 != k2.
```

This is exactly why I said in my previous comment:
> Note that we cannot simply check fisrt_full_key < key because this won't guarantee that first_key == key under the semantic of reverse iter.

If the UT passes with the current code, I think we lack a UT for this straight-forward case.

Reply:
I added some test for this case